Source/ASImageNode+AnimatedImage.mm

```diff
@@ -70,13 +71,20 @@ - (void)_locked_setAnimatedImage:(id <ASAnimatedImageProtocol>)animatedImage
       [self _locked_setShouldAnimate:YES];
     }
   } else {
-      // Clean up after ourselves.
-      self.contents = nil;
-      [self setCoverImage:nil];
+    // Clean up after ourselves.
+    shouldCleanup = YES;
   }
   
-  [self animatedImageSet:_animatedImage previousAnimatedImage:previousAnimatedImage];
+  {
+    // Unlock for cleanup and calling subclass hook
+    ASUnlockScope(self);
```

Comment:
Is this actually safe? We don't necessarily know what the calling scope is doing and therefore unlocking here could be dangerous? I know it's very unlikely (I can only think of convoluted examples)…

I'm not sure of a better solution though… If we do end up deciding this, we should at least add a comment that unlocking a lock someone else locked might cause us a bad time in the future.

Reply:
Agree. Another thing is that with a recursive lock, the lock may be acquired multiple times and so unlocking once is not enought.

Ideally, we have to either make this method `_u_` but it can be painful and may not be possible.

Another idea is dispatching to call `-animatedImageSet:previousAnimatedImage:` and the deallocation of previous image.